homeassistant/components/rfxtrx/binary_sensor.py

```diff
@@ -282,8 +273,6 @@ def apply_event(self, event):
 
         if self.hass:
             self.schedule_update_ha_state()
```

Comment:
Any reason that handle_event and apply_event are not 1 method. And could that method be `@callback`, and then we call `self.async_write_ha_state()` here?

Reply:
good idea

Reply:
I think we should wait with this change. The platform should be refactored to not access the entity from outside and the entity events should be dropped and should rework the sensor update first.

Reply:
It's not accessing from outside anymore since dispatcher changes went in. There is a global remaining. but the external access is gone.

Reply:
We're still accessing the entity from outside on line 124. That needs to go before we make more changes around the update.

Reply:
It's within the event loop and within the platform. Note it's also done on reaction to the exact same dispatcher event, that later on is linked to the exact same function.

Reply:
The entity api is only available after the entity has been added to home assistant. So we shouldn't access the entity from outside directly at all.

Reply:
I don't consider the change any worse than it was before (cleaner and simpler). But since it was initiated by @balloob, i'll wait for some agreement between you too on how to proceed. Need to wait for comments on the event stuff below anyway.

Reply:
It's worse since we need to change the api to async. We should wait with that.

Reply:
@MartinHjelmare would you be okey with something like this? I re-split the code between handling and application.
```patch
diff --git a/homeassistant/components/rfxtrx/sensor.py b/homeassistant/components/rfxtrx/sensor.py
index aef63ce2d2..bcbc190f02 100644
--- a/homeassistant/components/rfxtrx/sensor.py
+++ b/homeassistant/components/rfxtrx/sensor.py
@@ -95,7 +95,6 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
                 data_type = _data_type
                 break
         new_sensor = RfxtrxSensor(event, event.device, pkt_id, data_type)
-        new_sensor.apply_event(event)
         sub_sensors = {}
         sub_sensors[new_sensor.data_type] = new_sensor
         RFX_DEVICES[device_id] = sub_sensors
@@ -117,13 +116,16 @@ class RfxtrxSensor(Entity):
         self._unit_of_measurement = DATA_TYPES.get(data_type, "")
         self._unique_id = f"{slugify(device.type_string.lower())}_{slugify(device.id_string.lower())}_{slugify(data_type)}"
 
+        if event:
+            self._apply_event(event)
+
     async def async_added_to_hass(self):
         """Restore RFXtrx switch device state (ON/OFF)."""
         await super().async_added_to_hass()
 
         self.async_on_remove(
             self.hass.helpers.dispatcher.async_dispatcher_connect(
-                SIGNAL_EVENT, self.apply_event
+                SIGNAL_EVENT, self._handle_event
             )
         )
 
@@ -160,8 +162,11 @@ class RfxtrxSensor(Entity):
         """Return unique identifier of remote device."""
         return self._unique_id
 
+    def _apply_event(self, event):
+        self.event = event
+
     @callback
-    def apply_event(self, event):
+    def _handle_event(self, event):
         """Apply command from rfxtrx."""
         if not isinstance(event, SensorEvent):
             return
@@ -172,6 +177,5 @@ class RfxtrxSensor(Entity):
         if self.data_type not in event.values:
             return
 
-        self.event = event
-        if self.hass:
-            self.async_write_ha_state()
+        self._apply_event(event)
+        self.async_write_ha_state()
```

Reply:
I suppose that is pretty much going back to what @balloob wanted change. Feel like i'm stuck between a rock and a hard place.

Reply:
I think we can pass the initial event as a parameter to the entity and set the event to the event attribute in the ini method.

We can also directly set the event to the event attribute in `_handle_event`. Then we don't need `_apply_event` anymore and we won't access the entity from outside.

Reply:
We still want to parse the initial event though. Maybe do that in `async_added_to_hass`?

If we would remove the delayed reset of the binary sensor, which doesn't belong there, we could also parse the event in init method.

Reply:
Anyway, I think this whole change around event callback handling should be a separate PR. It's not a small one line fix. Please revert the changes around this for now.

Reply:
Okey. I've reverted back to the separate wrapper. Will work on change to this separate from this pull